cunumeric/module.py

```diff
@@ -6736,6 +6737,255 @@ def convolve(a: ndarray, v: ndarray, mode: ConvolveMode = "full") -> ndarray:
+        3. N arrays to specify the coordinates of the values along each
+           dimension of F. The length of the array must match the size of
+           the corresponding dimension
+        4. Any combination of N scalars/arrays with the meaning of 2. and 3.
+        If `axis` is given, the number of varargs must equal the number of
+        axes. Default: 1.
+    edge_order : {1, 2}, optional
+        Gradient is calculated using N-th order accurate differences
+        at the boundaries. Default: 1.
+        .. versionadded:: 1.9.1
+    axis : None or int or tuple of ints, optional
+        Gradient is calculated only along the given axis or axes
+        The default (axis = None) is to calculate the gradient for all the axes
+        of the input array. axis may be negative, in which case it counts from
+        the last to the first axis.
+        .. versionadded:: 1.11.0
+
+    Returns
+    -------
+    gradient : ndarray or list of ndarray
+        A list of ndarrays (or a single ndarray if there is only one dimension)
+        corresponding to the derivatives of f with respect to each dimension.
+        Each derivative has the same shape as f.
+
+    See Also
+    --------
+    numpy.gradient
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+
+    """
+    N = f.ndim  # number of dimensions
+
+    if axis is None:
+        axes = tuple(range(N))
+    elif isinstance(axis, collections.abc.Sequence):
+        axes = tuple(normalize_axis_index(a, N) for a in axis)
+    else:
+        axis = normalize_axis_index(axis, N)
+        axes = (axis,)
+
+    len_axes = len(axes)
+    if not varargs:
+        n = 0
+    else:
+        n = len(varargs)
+
+    if n == 0:
+        # no spacing argument - use 1 in all axes
+        dx = [asarray(1.0)] * len_axes
+    elif n == 1 and np.ndim(varargs[0]) == 0:
+        # single scalar for all axes
+        dx = list(asarray(varargs)) * len_axes
+    elif n == len_axes:
+        # scalar or 1d array for each axis
+        dx = list(asarray(v) for v in varargs)
+        for i, distances in enumerate(dx):
+            if distances.ndim == 0:
+                continue
+            elif distances.ndim != 1:
+                raise ValueError("distances must be either scalars or 1d")
+            if len(distances) != f.shape[axes[i]]:
+                raise ValueError(
+                    "when 1d, distances must match "
+                    "the length of the corresponding dimension"
+                )
+            if np.issubdtype(distances.dtype, np.integer):
+                # Convert numpy integer types to float64 to avoid modular
+                # arithmetic in np.diff(distances).
+                distances = distances.astype(np.float64)
+            diffx = diff(distances)
+            # if distances are constant reduce to the scalar case
+            # since it brings a consistent speedup
+            if (diffx == diffx[0]).all():
+                diffx = diffx[0]
```

Comment:
Let's remove this "optimization". It causes blocking in cuNumeric, and if the user wanted constant spacing they should have just provided a constant here.